### ISSUE TITLE
Boolean query with value boosting for terms

### ISSUE THREAD

**Issue by leehambley** [karmi/retire#422]
We're attempting to implement a search on documents including a _level_ field, this field will have a value matching `(?:low|medium|high)`, and should be sorted in preference of "best match". The semantics of best-match are that a query with _high_ should match documents stored with _high_ with a high score by preference, second _medium_, and lastly _low_. In the case that a search is made for a value matching _medium_, documents with high/low should match with (almost) equal prescience, behind pre-existing documents with the value _medium_.

A sample document may look like:

``` json
{
    "id": "ZpqsmszLeMdFZxhDrh2q", 
    "user_id": "kmGeJC1EEyvPRQyBfjyM",
    "level": "high", 
    "title": "Golf in Hammer Park"
}
```

Using `curl` we are able to generate correct results using the following query:

``` bash
curl -X GET "http://localhost:9200/announcements-test/_search?pretty=true" -d '{"query":{"bool":{"should":[{"term":{"level":{"value":"high","boost":3}}},{"term":{"level":{"value":"medium","boost":2}}},{"term":{"level":{"value":"low","boost":1}}}]}}}'
```

(Ignoring the obvious fact that we need to adapt the boost depending on the search term, and not boost with `3,2,1` mapped to `high, med, low` every time)

We've been trying to replicate this with the following Tire search:

``` ruby
    Tire.search(@@index_name) do
      query do
        boolean do
          should do
            term :level, { value: "high",   boost: 3}
          end
          should do
            term :level, { value: "medium", boost: 2}
          end
          should do
            term :level, { value: "low",    boost: 1}
          end
        end
      end
    end
```

Tire generates the following (`#to_curl`):

``` bash
curl -X GET "http://localhost:9200/announcements-test/_search?pretty=true" -d '{"query":{"bool":{"should":[{"term":{"level":{"term":{"value":"high","boost":3}}}},{"term":{"level":{"term":{"value":"medium","boost":2}}}},{"term":{"level":{"term":{"value":"low","boost":1}}}}]}}}'
```

The rabbit hole leads to [`search/query.rb#11`](https://github.com/karmi/tire/blob/master/lib/tire/search/query.rb#L11) where a `:term =>` key is added to whatever options are passed (admittedly, this is probably correct in 99.9% of cases)

_My question_ thus is whether we should send a pull request for the correct (more flexible behaviour), if there is a different way you prefer to recommend to get the same result, or whether we could infact simply formulate the JSON query ourselves, and use the Tire engine to perform the query.

The last option, I had thought was possible, but I haven't found any documentation to that effect, although I'm on the wrong end of a 12 hour day, so maybe I'll find that later!

**Comment by pex** [comment]
same issue here :scream:

**Comment by leehambley** [comment]
https://github.com/karmi/tire/pull/423

**Comment by karmi** [comment]
Lee, my apologies for the delay -- thanks!, that of course makes perfect sense.

Fixed some minor issue with a Ruby 1.9-styled Hash, and also added `to_hash` so you can use Hash-like objects (Hashr, Mash, etc) as well.

**Comment by leehambley** [comment]
Whoohoo, thanks :-) :+1:
